airsonic-main/src/main/java/org/airsonic/player/service/InternetRadioService.java

```diff
@@ -0,0 +1,112 @@
+package org.airsonic.player.service;
+
+import chameleon.playlist.*;
+import org.airsonic.player.domain.InternetRadio;
+import org.airsonic.player.domain.InternetRadioSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import java.net.URL;
+import java.util.*;
+
+@Service
+public class InternetRadioService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(InternetRadioService.class);
+
+    private Map<Integer, List<InternetRadioSource>> cachedSources;
+
+    public InternetRadioService() {
+        this.cachedSources = new HashMap<>();
+    }
+
+    public void clearInternetRadioSourceCache() {
+        cachedSources.clear();
+    }
+
+    public List<InternetRadioSource> getInternetRadioSources(InternetRadio radio) throws Exception {
+        List<InternetRadioSource> sources;
+        if (cachedSources.containsKey(radio.getId())) {
+            LOG.info("Got cached sources for internet radio {}!", radio.getStreamUrl());
+            sources = cachedSources.get(radio.getId());
+        } else {
+            LOG.info("Retrieving sources for internet radio {}...", radio.getStreamUrl());
+            sources = retrieveInternetRadioSources(radio);
+            if (sources.isEmpty()) {
+                LOG.warn("No entries found when parsing external playlist.");
+            }
+            LOG.info("Finished parsing playlist for internet radio {}, got {} sources.", radio.getStreamUrl(), sources.size());
+            cachedSources.put(radio.getId(), sources);
+        }
+        return sources;
+    }
+
+    private List<InternetRadioSource> retrieveInternetRadioSources(InternetRadio radio) throws Exception {
+
+        // Retrieve radio playlist and parse it
+        URL playlistUrl = new URL(radio.getStreamUrl());
+        SpecificPlaylist inputPlaylist = null;
+        try {
+            LOG.debug("Parsing internet radio playlist at {}...", playlistUrl.toString());
+            inputPlaylist = SpecificPlaylistFactory.getInstance().readFrom(playlistUrl);
+        } catch (Exception e) {
+            LOG.error("Unable to parse internet radio playlist: {}", playlistUrl.toString(), e);
+            throw e;
```

Comment:
Don't we want to simply return an empty list, instead of yielding the exception?

Reply:
Hmm, I haven't managed yet to work on forwarding parsing errors to the UI, which is why I defaulted to throwing the exception. Any ideas on how to do this properly?

Reply:
I guess that we should add an helper to do this then, in an other PR.
I guess yielding an exception is fine for now :)